docroot/modules/custom/prisoner_hub_entity_access/src/EventSubscriber/QueryAccessSubscriber.php

```diff
@@ -0,0 +1,178 @@
+   *
+   * @var RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * The prison category field name.
+   *
+   * @var String
+   */
+  protected $prisonCategoryFieldName;
+
+  /**
+   * The prison category field name.
+   *
+   * @var String
+   */
+  protected $prisonFieldName;
+
+  public function __construct(EntityFieldManagerInterface $entity_field_manager, RouteMatchInterface $route_match, string $prison_field_name, string $prison_category_field_name) {
+    $this->entityFieldManager = $entity_field_manager;
+    $this->routeMatch = $route_match;
+    $this->prisonFieldName = $prison_field_name;
+    $this->prisonCategoryFieldName = $prison_category_field_name;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents() {
+    // TODO: Do we want to run this for Taxonomy Terms?  Or other entity types?
+    $events['entity.query_access.node'] = ['entityQueryAccessPrisonCategories'];
+
+    return $events;
+  }
+
+  /**
+   * This method is called when the entity.query_access.prison_categories is dispatched.
+   *
+   * @param \Symfony\Component\EventDispatcher\Event $event
+   *   The dispatched event.
+   */
+  public function entityQueryAccessPrisonCategories(QueryAccessEvent $event) {
+    $operations = ['view', 'view label'];
+
+    /* @var \Drupal\taxonomy\TermInterface $current_prison */
+    $current_prison = $this->routeMatch->getParameter('prison');
+    if (!in_array($event->getOperation(), $operations) || !$current_prison) {
+      return;
+    }
+
+    $conditions = $event->getConditions();
+    $conditions->alwaysFalse(FALSE);
+
+    $current_prison_condition_group = $this->getPrisonConditionGroup($current_prison);
+    $prison_categories_condition_group = $this->getPrisonCategoriesConditionGroup($current_prison);
+
+    // Only create an OR condition group if a prison categories condition is to
+    // be added.
+    if ($prison_categories_condition_group) {
+      $condition_group = new ConditionGroup('OR');
+      $condition_group->addCondition($current_prison_condition_group);
+      $condition_group->addCondition($prison_categories_condition_group);
+      $conditions->addCondition($condition_group);
+    }
+    else {
+      $conditions->addCondition($current_prison_condition_group);
+    }
+  }
+
+  /**
+   * Get the condition group for content tagged with the current prison.
+   *
+   * @param \Drupal\taxonomy\TermInterface $current_prison
+   *   The current prison taxonomy term (loaded via the url).
+   *
+   * @return \Drupal\entity\QueryAccess\ConditionGroup
+   *   The condition group, to be added to the entity query.
+   */
+  protected function getPrisonConditionGroup(TermInterface $current_prison) {
+    $condition_group = new ConditionGroup('OR');
+    $condition_group->addCondition($this->prisonFieldName, $current_prison->id());
+    $bundles = $this->getFieldBundles('node', $this->prisonFieldName);
+    $condition_group->addCondition('type', $bundles, 'NOT IN');
```

Comment:
Sorry, noob question: what are `FieldBundles` and what does this `OR` group do?

Reply:
So that we can have the flexibility to have some content types _without_ the prison category fields.  We first get all the bundles (i.e. content types) that have the field enabled, and then do we a `OR NOT IN` on those bundles.  I.e. we are only applying the rule to bundles that have the field, on those without we always show the content.

This was the only way I could think of to represent this in a query (as the query at this point could span over multiple bundles).  
Maybe it's worth I add some further documentation on this, as it's probably not the most straightforward thing to grasp?

Reply:
Added some inline comments in the code for this.